lib/ex_doc/formatter/html/search_items.ex

```diff
@@ -2,64 +2,85 @@ defmodule ExDoc.Formatter.HTML.SearchItems do
   alias ExDoc.Formatter.HTML
 
   def create(nodes, extras) do
     items = Enum.flat_map(nodes, &module/1) ++ Enum.flat_map(extras, &extra/1)
     ["searchNodes=" | ExDoc.Utils.to_json(items)]
   end
 
-  @h2_split_regex ~r/<h2\b.*?>/
-  @header_body_regex ~r/(?<header>.+)<\/h2>(?<body>.*)/s
-  defp extra(%{id: id, title: title, content: content}) do
-    [intro | sections] = Regex.split(@h2_split_regex, content)
-    intro_json_item = encode("#{id}.html", title, :extras, intro)
+  defp extra(map) do
+    [intro | sections] = Regex.split(~r/## (?<header>\b.+)/, map.source, include_captures: true)
+
+    sections = Enum.chunk_every(sections, 2)
+
+    intro_json_item =
+      encode(
+        "#{map.id}.html",
+        map.title,
+        :extras,
+        String.trim(intro)
+      )
 
     section_json_items =
-      sections
-      |> Enum.map(&Regex.named_captures(@header_body_regex, &1))
-      |> Enum.map(&extra_section(title, &1["header"], &1["body"], id))
+      for [header, body] <- sections do
+        "## " <> header = header
+
+        encode(
+          "#{map.id}.html##{HTML.text_to_id(header)}",
+          "#{map.title} - " <> clean_doc(header),
+          :extras,
+          body
+        )
+      end
 
     [intro_json_item | section_json_items]
   end
 
-  defp extra_section(title, header, body, id) do
-    header = HTML.strip_tags(header)
+  defp module(%ExDoc.ModuleNode{} = node) do
+    source_doc =
+      node.doc_format == "text/markdown" && is_map(node.source_doc) && node.source_doc["en"]
 
-    encode(
-      "#{id}.html##{HTML.text_to_id(header)}",
-      "#{title} - #{header}",
-      :extras,
-      body
-    )
-  end
+    module =
+      encode(
+        "#{node.id}.html",
+        node.id,
+        node.type,
+        source_doc || node.rendered_doc
+      )
 
-  defp module(node = %ExDoc.ModuleNode{id: id, type: type, rendered_doc: doc}) do
-    module = encode("#{id}.html", id, type, doc)
-    functions = Enum.map(node.docs, &node_child(&1, id))
-    types = Enum.map(node.typespecs, &node_child(&1, id))
+    functions = Enum.map(node.docs, &node_child(&1, node))
+    types = Enum.map(node.typespecs, &node_child(&1, node))
     [module] ++ functions ++ types
   end
 
-  defp node_child(node, module) do
+  defp node_child(node, module_node) do
+    source_doc =
+      module_node.doc_format == "text/markdown" && is_map(node.source_doc) &&
+        node.source_doc["en"]
+
     encode(
-      "#{module}.html##{node.id}",
-      "#{module}.#{node.name}/#{node.arity}",
+      "#{module_node.id}.html##{node.id}",
+      "#{module_node.id}.#{node.name}/#{node.arity}",
       node.type,
-      node.rendered_doc
+      source_doc || node.rendered_doc
     )
   end
 
   defp encode(ref, title, type, doc) do
-    %{ref: URI.encode(ref), title: title, type: type, doc: clean_doc(doc)}
+    %{
+      ref: URI.encode(ref),
+      title: title,
+      type: type,
+      doc: clean_doc(doc)
```

Comment:
We should only clean this doc for HTML. For Markdown, let's keep everything as is?

Reply:
Maybe we should even move this to the language bits? They both have to return the "html_doc" and the "text_doc/search_doc"?

Reply:
it's rare but markdown may contain html tags and we shouldn't render them on the search page.

Reply:
Hrm... I am worried about removing something we should not when we strip the white-space. Maybe we strip only the tags for Markdown?

Reply:
Agreed, for markdown we just clean tags from markdown, for Erlang we strip tags & clean whitespace from _html_.

I'll look into moving this behaviour to the language contract later, after removing duplication from sidebar/search generation.